src/esockd_connection_sup.erl

```diff
@@ -108,37 +114,42 @@ init([Opts, MFA]) ->
     MaxConns = get_value(max_connections, Opts, ?DEFAULT_MAX_CONNS),
     RawRules = get_value(access_rules, Opts, [{allow, all}]),
     AccessRules = [esockd_access:compile(Rule) || Rule <- RawRules],
-    {ok, #state{curr_connections = 0, max_connections  = MaxConns,
-                access_rules = AccessRules, shutdown = Shutdown, mfargs = MFA}}.
+    {ok, #state{curr_connections = #{},
+                max_connections  = MaxConns,
+                access_rules     = AccessRules,
+                shutdown         = Shutdown,
+                mfargs           = MFA}}.
 
 handle_call({start_connection, _Sock}, _From,
-            State = #state{curr_connections = CurrConns, max_connections  = MaxConns})
-    when CurrConns >= MaxConns ->
+            State = #state{curr_connections = Conns, max_connections = MaxConns})
+    when map_size(Conns) >= MaxConns ->
     {reply, {error, maxlimit}, State};
 
-handle_call({start_connection, Sock}, _From, State = #state{mfargs = MFA, curr_connections = Count, access_rules = Rules}) ->
+handle_call({start_connection, Sock}, _From,
+            State = #state{curr_connections = Conns, access_rules = Rules, mfargs = MFA}) ->
     case esockd_transport:peername(Sock) of
         {ok, {Addr, _Port}} ->
             case allowed(Addr, Rules) of
-                true  -> case catch start_connection_proc(MFA, Sock) of
-                             {ok, Pid} when is_pid(Pid) ->
-                                 put(Pid, true),
-                                 {reply, {ok, Pid}, State#state{curr_connections = Count+1}};
-                             ignore ->
-                                 {reply, ignore, State};
-                             {error, Reason} ->
-                                 {reply, {error, Reason}, State};
-                             What ->
-                                {reply, {error, What}, State}
-                         end;
-                false -> {reply, {error, forbidden}, State}
+                true ->
+                    case catch start_connection_proc(MFA, Sock) of
```

Comment:
direct `catch` of an expression is considered deprecated.
always use try catch like below instead:
```
try
  expr
catch _ : _ ->
end.```